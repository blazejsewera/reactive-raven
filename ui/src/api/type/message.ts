export type ApiMessage = {
  username: string
  timestamp: string
  title: string
  subtitle?: string
  body?: string
  id?: string
}

// example:
// { "username": "Test", "timestamp": "a", "title": "Test title", "subtitle": "Test subtitle", "body": "Test body.", "id": "1234" }
