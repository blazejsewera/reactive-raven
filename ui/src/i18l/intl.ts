/**
 * Intl is an interface for an internationalization (i18l) tool
 * that can provide translations for page.
 */
export interface Intl {
  getTranslation: (text: string) => string
}

export const intl: Intl = { getTranslation: text => text }
