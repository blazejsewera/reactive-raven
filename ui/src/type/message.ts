export type Message = {
  username: string
  timestamp: string
  relativeTime?: string
  userImgUri?: string
  title: string
  subtitle?: string
  body?: string
  id?: string
}
