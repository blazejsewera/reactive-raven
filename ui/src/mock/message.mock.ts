import { Message } from '../type/message'
import { fiveSentenceLoremIpsum } from './asset/text/lipsum'
import githubIcon from './asset/icon/github-icon.svg'

const body = `This is a simple test message.

It spans over multiple lines.`

export const full: Message = {
  username: 'Primary',
  title: 'Test title',
  subtitle: '— primary',
  body,
  timestamp: '2 hours ago',
  id: '0',
}

export const fullWithImage: Message = {
  ...full,
  id: '1',
  userImgUri: githubIcon,
}

export const fullWithLoremIpsum: Message = {
  ...full,
  id: '2',
  body: fiveSentenceLoremIpsum,
}

export const partial: Message = {
  username: full.username,
  title: full.title,
  body: full.body,
  timestamp: full.timestamp,
  id: '3',
}

export const minimal: Message = {
  username: full.username,
  title: full.title,
  timestamp: full.timestamp,
  id: '4',
}

export const anotherUser: Message = {
  username: 'Other',
  title: 'Other title',
  subtitle: '— other',
  body,
  timestamp: '1 hour ago',
  id: '5',
}
