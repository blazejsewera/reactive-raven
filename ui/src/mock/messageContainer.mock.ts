import type { MessageContainerHandlers } from '../type/handler'

export const handlers: MessageContainerHandlers = {
  onClear: () => {},
}
