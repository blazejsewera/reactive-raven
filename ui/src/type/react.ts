import type { ReactElement, ReactNode } from "react";

type EmptyObject = Record<string, never>;

/**
 * Default children type in React. Every renderable type is included.
 */
export type DefaultChildren = ReactNode;

/**
 * No props, a.k.a. () => </> component
 */
export type EmptyProps = EmptyObject;

/**
 * No React Context
 */
export type EmptyContext = EmptyObject;

/**
 * Functional Component (strongly typed)
 */
export interface FC<P = EmptyProps, Ctx = EmptyContext> {
  (props: P, context?: Ctx): ReactElement | null;
  displayName?: string;
}
