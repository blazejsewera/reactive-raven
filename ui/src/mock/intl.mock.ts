import { Intl } from '../i18l/intl'

export const intlMock: Intl = { getTranslation: text => text }
