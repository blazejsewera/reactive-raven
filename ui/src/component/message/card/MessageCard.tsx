import * as React from 'react'
import { FC } from '../../../type/react'
import { Message } from '../../../type/message'
import { Intl } from '../../../i18l/intl'
import { cx } from '../../../util/classname/cx'
import { Header } from './section/Header'
import { Body } from './section/Body'
import { Footer } from './section/Footer'

export interface MessageCardProps {
  message: Message
  intl: Intl
}

export const MessageCard: FC<MessageCardProps> = ({ message, intl }) => {
  const { username, userImgUri: appImgUri, title, subtitle, body, timestamp, relativeTime } = message

  return (
    <div className={cx('rounded-3xl', 'bg-white', 'dark:bg-gray-800', 'w-80', 'p-5', 'shadow-lg', 'relative')}>
      <Header {...{ username, appImgUri, title, subtitle }} />
      <Body intl={intl}>{body ?? ''}</Body>
      <Footer {...{ username, timestamp, relativeTime }} />
    </div>
  )
}
