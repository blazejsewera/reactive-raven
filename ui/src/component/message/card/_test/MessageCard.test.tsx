import * as React from 'react'
import { create as render } from 'react-test-renderer'
import type { Message } from '../../../../type/message'
import { MessageCard } from '../MessageCard'
import { intlMock } from '../../../../mock/intl.mock'
import { full, fullWithImage, fullWithLoremIpsum, minimal, partial } from '../../../../mock/message.mock'

describe('MessageCard component', () => {
  // given
  const intl = intlMock
  const testMessageCard = (message: Message) => {
    const tree = render(<MessageCard {...{ message, intl }} />).toJSON()
    expect(tree).toMatchSnapshot()
  }

  it('renders correctly with full message', () => {
    testMessageCard(full)
  })
  it('renders correctly with full with image message', () => {
    testMessageCard(fullWithImage)
  })
  it('renders correctly with full with very long text message', () => {
    testMessageCard(fullWithLoremIpsum)
  })
  it('renders correctly with partial message', () => {
    testMessageCard(partial)
  })
  it('renders correctly with minimal message', () => {
    testMessageCard(minimal)
  })
})
