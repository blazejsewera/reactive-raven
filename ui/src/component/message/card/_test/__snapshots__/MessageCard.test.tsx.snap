// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageCard component renders correctly with full message 1`] = `
<div
  className="rounded-3xl bg-white dark:bg-gray-800 w-80 p-5 shadow-lg relative"
>
  <div
    className="flex"
  >
    <div
      className="h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default"
    >
      <svg
        viewBox="0 0 1 1"
      >
        <rect
          fill="#000"
          height={1}
          width={1}
        />
        <text
          dy="0.178"
          fill="#FFF"
          style={
            Object {
              "fontFamily": "Inter",
              "fontSize": 0.5,
              "fontWeight": "bold",
            }
          }
          textAnchor="middle"
          x={0.5}
          y={0.5}
        >
          Pr
        </text>
      </svg>
    </div>
    <div
      className="my-auto inline-block ml-5"
    >
      <p
        className="text-lg font-bold mb-0 leading-5 text-gray-800 dark:text-gray-300"
      >
        Test title
      </p>
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        — primary
      </p>
    </div>
  </div>
  <p
    className="whitespace-pre-line text-sm leading-4 my-3 text-gray-800 dark:text-gray-300"
  >
    <span>
      This is a simple test message.

It spans over multiple lines.
    </span>
  </p>
  <div
    className="flex"
  >
    <div
      className="inline-block"
    >
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        — by 
        Primary
      </p>
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        2 hours ago
      </p>
    </div>
  </div>
</div>
`;

exports[`MessageCard component renders correctly with full with image message 1`] = `
<div
  className="rounded-3xl bg-white dark:bg-gray-800 w-80 p-5 shadow-lg relative"
>
  <div
    className="flex"
  >
    <div
      className="h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default border dark:bg-gray-300 dark:border-gray-300"
    >
      <img
        src="test-file-stub"
      />
    </div>
    <div
      className="my-auto inline-block ml-5"
    >
      <p
        className="text-lg font-bold mb-0 leading-5 text-gray-800 dark:text-gray-300"
      >
        Test title
      </p>
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        — primary
      </p>
    </div>
  </div>
  <p
    className="whitespace-pre-line text-sm leading-4 my-3 text-gray-800 dark:text-gray-300"
  >
    <span>
      This is a simple test message.

It spans over multiple lines.
    </span>
  </p>
  <div
    className="flex"
  >
    <div
      className="inline-block"
    >
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        — by 
        Primary
      </p>
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        2 hours ago
      </p>
    </div>
  </div>
</div>
`;

exports[`MessageCard component renders correctly with full with very long text message 1`] = `
<div
  className="rounded-3xl bg-white dark:bg-gray-800 w-80 p-5 shadow-lg relative"
>
  <div
    className="flex"
  >
    <div
      className="h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default"
    >
      <svg
        viewBox="0 0 1 1"
      >
        <rect
          fill="#000"
          height={1}
          width={1}
        />
        <text
          dy="0.178"
          fill="#FFF"
          style={
            Object {
              "fontFamily": "Inter",
              "fontSize": 0.5,
              "fontWeight": "bold",
            }
          }
          textAnchor="middle"
          x={0.5}
          y={0.5}
        >
          Pr
        </text>
      </svg>
    </div>
    <div
      className="my-auto inline-block ml-5"
    >
      <p
        className="text-lg font-bold mb-0 leading-5 text-gray-800 dark:text-gray-300"
      >
        Test title
      </p>
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        — primary
      </p>
    </div>
  </div>
  <p
    className="whitespace-pre-line text-sm leading-4 my-3 text-gray-800 dark:text-gray-300"
  >
    <span>
      Explicabo ipsum iusto qui porro iste possimus ad voluptatem. Nam maxime voluptatem nostrum deleniti saepe. Officia dolor velit sunt velit placeat amet
      <span
        aria-label="show more"
        className="text-blue-500 dark:text-blue-400 cursor-pointer"
        onClick={[Function]}
      >
        [...]
      </span>
    </span>
  </p>
  <div
    className="flex"
  >
    <div
      className="inline-block"
    >
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        — by 
        Primary
      </p>
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        2 hours ago
      </p>
    </div>
  </div>
</div>
`;

exports[`MessageCard component renders correctly with minimal message 1`] = `
<div
  className="rounded-3xl bg-white dark:bg-gray-800 w-80 p-5 shadow-lg relative"
>
  <div
    className="flex"
  >
    <div
      className="h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default"
    >
      <svg
        viewBox="0 0 1 1"
      >
        <rect
          fill="#000"
          height={1}
          width={1}
        />
        <text
          dy="0.178"
          fill="#FFF"
          style={
            Object {
              "fontFamily": "Inter",
              "fontSize": 0.5,
              "fontWeight": "bold",
            }
          }
          textAnchor="middle"
          x={0.5}
          y={0.5}
        >
          Pr
        </text>
      </svg>
    </div>
    <div
      className="my-auto inline-block ml-5"
    >
      <p
        className="text-lg font-bold mb-0 leading-5 text-gray-800 dark:text-gray-300"
      >
        Test title
      </p>
    </div>
  </div>
  <p
    className="whitespace-pre-line text-sm leading-4 my-3 text-gray-800 dark:text-gray-300"
  >
    <span>
      
    </span>
  </p>
  <div
    className="flex"
  >
    <div
      className="inline-block"
    >
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        — by 
        Primary
      </p>
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        2 hours ago
      </p>
    </div>
  </div>
</div>
`;

exports[`MessageCard component renders correctly with partial message 1`] = `
<div
  className="rounded-3xl bg-white dark:bg-gray-800 w-80 p-5 shadow-lg relative"
>
  <div
    className="flex"
  >
    <div
      className="h-12 w-12 rounded-full overflow-clip overflow-hidden inline-block cursor-default"
    >
      <svg
        viewBox="0 0 1 1"
      >
        <rect
          fill="#000"
          height={1}
          width={1}
        />
        <text
          dy="0.178"
          fill="#FFF"
          style={
            Object {
              "fontFamily": "Inter",
              "fontSize": 0.5,
              "fontWeight": "bold",
            }
          }
          textAnchor="middle"
          x={0.5}
          y={0.5}
        >
          Pr
        </text>
      </svg>
    </div>
    <div
      className="my-auto inline-block ml-5"
    >
      <p
        className="text-lg font-bold mb-0 leading-5 text-gray-800 dark:text-gray-300"
      >
        Test title
      </p>
    </div>
  </div>
  <p
    className="whitespace-pre-line text-sm leading-4 my-3 text-gray-800 dark:text-gray-300"
  >
    <span>
      This is a simple test message.

It spans over multiple lines.
    </span>
  </p>
  <div
    className="flex"
  >
    <div
      className="inline-block"
    >
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        — by 
        Primary
      </p>
      <p
        className="text-xs font-bold text-gray-500 dark:text-gray-400"
      >
        2 hours ago
      </p>
    </div>
  </div>
</div>
`;
